test/test_broker.hpp

```diff
@@ -1057,6 +1098,7 @@ class test_broker {
 
         MQTT_NS::buffer client_id;
         con_sp_t con;
+        std::map<MQTT_NS::buffer, MQTT_NS::qos> subs;
```

Comment:
I'm very concerned here.

Not only are you adding each incoming subscription to the existing "subs_" map.
But you're also adding a new "subs_map"

*and* adding a map of subscription information to each session object (one for each client).

There should only be a single place that any particular piece of data lives, and if you need to access that data in more than one way, all access methods should ultimately point to the same underlying storage location. Exceptions to this rule exist, but they are fairly rare.

Reply:
The subs_ can actually be removed. Please note that the subs_map does not store the actual subscription data, rather, builds an index separate from the data. This is also what is done in the multi_index btw, the indices are derived from the data actually stored in the struct. 

What could be done is the following: 
Store a subscription in the subs_map, and store the key of the subscription within the session. The key would then be: <parent, last path token, Value>. Once the session is deleted, you remove all keys from the subscription database. 

The session would then only need to store the list of keys it has generated in the subscription index, to be able to clear them once the session is clear.

Reply:
The issue with the key of  the subscription map is that the key is not only the subscription path. You can have multiple sessions with the same subscription, so the key is actually <subscription, session>. Maybe it is actually better to be able to lookup a subscription based on the subscription and based on the session id. So if we use a multiindex as base, we can realise this.